### PR TITLE
fix docs+examples, work without reset, fixes

### PR DESCRIPTION
**Pull request by burrbull**
r? @therealprof 

Add more docs as @Disasm proposed.

Restore `unsafe` on `W::bits()` I somewhere lost.

Don't generate `write` if `reset_value` is not present. #258

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -256,26 +262,30 @@
 //! }
 //! ```
 //!
+//! ## `reset`
+//!
+//! `ResetValue` trait provides `reset_value` which returns the value of the `CR2`
+//! register after a reset. This value can be used to modify the
+//! writable bitfields of the `CR2` register or reset it in initial state.
```
Comment (therealprof) [inline src/lib.rs]:
```suggestion
//! writable bitfields of the `CR2` register or reset it to its initial state.
```

File: src/lib.rs [hunk 2]
```diff
@@ -256,26 +262,30 @@
 //! }
 //! ```
 //!
+//! ## `reset`
+//!
+//! `ResetValue` trait provides `reset_value` which returns the value of the `CR2`
```
Comment (therealprof) [inline src/lib.rs]:
```suggestion
//! The `ResetValue` trait provides `reset_value` which returns the value of the `CR2`
```

File: src/generate/generic.rs [hunk 3]
```diff
@@ -165,16 +208,18 @@ impl<FI> R<bool, FI> {
 }
 
 ///Register writer
+///
+///It is used as argument of closures in [`write`](Reg::write) and [`modify`](Reg::modify) methods of register
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
///Used as an argument to the closures in the [`write`](Reg::write) and [`modify`](Reg::modify) methods of the register
```

File: src/generate/generic.rs [hunk 4]
```diff
@@ -112,6 +151,9 @@ where
 }
 
 ///Register/field reader
+///
+///Result of call [`read`](Reg::read) method of register.
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
///Result of the [`read`](Reg::read) method of a register.
```

File: src/generate/generic.rs [hunk 5]
```diff
@@ -112,6 +151,9 @@ where
...
+///Also it can be used in [`modify`](Reg::read) method
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
///Also it can be used in the [`modify`](Reg::read) method
```

File: src/generate/generic.rs [hunk 6]
```diff
@@ -100,7 +125,21 @@ where
 {
     ///Modifies the contents of the register
     ///
-    ///See [modifying](https://rust-embedded.github.io/book/start/registers.html#modifying) in book.
+    ///Change only part of register:
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///E.g. to do a read-modify-write sequence to change parts of a register:
```

File: src/generate/generic.rs [hunk 7]
```diff
@@ -100,7 +125,21 @@ where
...
+    ///```ignore
+    ///periph.reg.modify(|r, w| unsafe { w.bits(
+    ///   r.bits() | 3
+    ///) });
+    ///```
+    ///or
+    ///```ignore
+    ///periph.reg.modify(|_, w| w
+    ///    .field1() .bits(newfield1bits)
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field1().bits(newfield1bits)
```

File: src/generate/generic.rs [hunk 8]
```diff
@@ -100,7 +125,21 @@ where
...
+    ///    .field2() .set_bit()
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field2().set_bit()
```

File: src/generate/generic.rs [hunk 9]
```diff
@@ -100,7 +125,21 @@ where
...
+    ///    .field3() .variant(VARIANT)
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field3().variant(VARIANT)
```

File: src/generate/generic.rs [hunk 10]
```diff
@@ -68,7 +79,19 @@ where
...
     ///
-    ///See [writing](https://rust-embedded.github.io/book/start/registers.html#writing) in book.
+    ///You can write raw bits into register:
+    ///```ignore
+    ///periph.reg.write(|w| unsafe { w.bits(rawbits) });
+    ///```
+    ///or write only fields you need:
+    ///```ignore
+    ///periph.reg.write(|w| w
+    ///    .field1() .bits(newfield1bits)
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field1().bits(newfield1bits)
```

File: src/generate/generic.rs [hunk 11]
```diff
@@ -68,7 +79,19 @@ where
...
+    ///    .field2() .set_bit()
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field2().set_bit()
```

File: src/generate/generic.rs [hunk 12]
```diff
@@ -68,7 +79,19 @@ where
...
+    ///    .field3() .variant(VARIANT)
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///    .field3().variant(VARIANT)
```

File: src/generate/generic.rs [hunk 13]
```diff
@@ -68,7 +79,19 @@ where
 {
     ///Writes bits to `Writable` register
     ///
-    ///See [writing](https://rust-embedded.github.io/book/start/registers.html#writing) in book.
+    ///You can write raw bits into register:
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///You can write raw bits into a register:
```

File: src/generate/generic.rs [hunk 14]
```diff
@@ -68,7 +79,19 @@ where
...
+    ///```ignore
+    ///periph.reg.write(|w| unsafe { w.bits(rawbits) });
+    ///```
+    ///or write only fields you need:
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///or write only the fields you need:
```

File: src/generate/generic.rs [hunk 15]
```diff
@@ -42,7 +42,16 @@ where
 {
     ///Reads the contents of `Readable` register
     ///
-    ///See [reading](https://rust-embedded.github.io/book/start/registers.html#reading) in book.
+    ///You can read contents of register in such way:
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///You can read the contents of a register in such way:
```

File: src/generate/generic.rs [hunk 16]
```diff
@@ -42,7 +42,16 @@ where
...
+    ///```ignore
+    ///let bits = periph.reg.read().bits();
+    ///```
+    ///or get contents of particular field of register.
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///or get the content of a particular field of a register.
```

File: src/generate/generic.rs [hunk 17]
```diff
@@ -55,6 +64,8 @@ where
     U: Copy,
 {
     ///Writes the reset value to `Writable` register
+    ///
+    ///After call this method register takes the initial state
```
Comment (therealprof) [inline src/generate/generic.rs]:
```suggestion
    ///Resets the register to its initial state
```

**Comment by burrbull** [comment]
Done. Rebased.

**Pull request merged**